src/diffusers/models/unet_2d_blocks.py

```diff
@@ -762,6 +790,100 @@ def forward(self, hidden_states, temb=None, upsample_size=None):
         return hidden_states, output_states
 
 
+class AttnDownsampleBlock2D(nn.Module):
```

Comment:
Sorry if I asked previously but do we need a new unet block? Which of the existing ones are the closest and why can't we use them?

Reply:
@ayushtues 

I believe the reason is because the [down](https://github.com/openai/consistency_models/blob/main/cm/unet.py#L596)/[upsample](https://github.com/openai/consistency_models/blob/main/cm/unet.py#L676) blocks in the original code and checkpoints have a [[ResNet block](https://github.com/openai/consistency_models/blob/main/cm/unet.py#L143) => [Attention block](https://github.com/openai/consistency_models/blob/main/cm/unet.py#L259)] x N => ResNet block downsampler structure, which doesn't correspond to any existing `diffusers` unet blocks.

The most similar existing unet blocks to my knowledge are:

- [`AttnDownBlock2D`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py#L710)/[`AttnUpBlock2D`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py#L1806) which have structure [ResNet block => Attention block] x N => downsampler but use a convolutional down/upsampler ([`Downsample2D`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/resnet.py#L175)/[`Upsample2D`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/resnet.py#L103)) instead of a ResNet downsampler.
- [`SimpleCrossAttnDownBlock2D`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py#L1441)/[`SimpleCrossAttnUpBlock2D`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py#L2552), which have structure [ResNet block => Attention block] x N => ResNet downsampler, but the Attention block is configured to use cross-attention instead of self-attention like the original implementation.

See also https://github.com/dg845/diffusers/pull/1#issuecomment-1562574268 and the preceding discussion.

Reply:
Thanks for the great explanation here. I understand the problem, but here I think it would be better if you could try to adapt `AttnDownBlock2D`: https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py#L710 
and 
https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py#L1806

so that they can use both a ResNet and a Conv up/down sampler. 

Happy to help with this as well if needed